ARCHITECTURE.md

```diff
@@ -0,0 +1,221 @@
+
+CodeRenderer
+------------
+
+CodeRenderer is a standalone class that takes the result of the CodeGenerator class and constructs files.
+There are 3 files constructed:
+
+- The C file with the decoding/encoding functions.
+- The H file with the public API to some functions in the C file.
+- The H file with all the struct definitions (the type file). If both decoding and encoding files are generated for the same CDDL, they can share the same type file.
+
+CodeRenderer conducts some pruning and deduplication of the list of types and functions received from CodeGenerator.
+
+
+Architecture of the generated C code
+====================================
+
+In the generated C file, each type from the CDDL file gets its own decoding/encoding function, unless they are trivial types, like `Foo = uint`.
+These functions are all `static`.
+In addition, all types in `entry_types` get public wrapper functions.
+
+All decoding/encoding functions operate on a state variable of type `cbor_state_t` which keeps track of:
+
+- The current position in the payload, and the end of the payload.
+- The current position in a list or map, and the maximum expected number of elements.
+- A list of backup states, used for saving states so they can be restored if decoding/encoding fails while processing an optional element.
+
+Each function returns a `bool` indicating whether it was successful at decoding/encoding.
+In most cases, a failure in one function will result in a failure of the whole operation.
+
+However, in the following scenarios, a failure is fine because we don't know ahead of time whether the object will be found or not:
+
+- An object with unknown number of repetitions (`minQ` and `maxQ` are not the same).
+- `"UNION"`s, since only one of the children should be present.
+
+In these cases, the code attempts to decode, and if it fails, it restores the state to before the attempt, and tries decoding the next candidate type.
+
+All generated functions take the form of a single if statement.
+This if statement performs boolean algebra on statements depending on the children (typically only container types get a generated function).
+The assignment of values in the structs mostly happens in the non-generated code.
+The generated code mostly combines and validates calls into the non-generated code or other generated functions.
+
+All functions (generated and not) have the same API structure: `bool <name>(cbor_state_t *state, <type> *result)`.
+The number of arguments is kept to a minimum to reduce code size.
+
+The exceptions to the API structure are `multi_decode`/`multi_encode` and `present_decode`/`present_encode`.
+These functions accept function pointers with the above API and runs them multiple times.
+When this happens, the functions pointers are cast to a generic function pointer type, and processed without knowledge of the type.
```

Comment:
```suggestion
When this happens, the functions' pointers are cast to a generic function pointer type, and processed without knowledge of the type.
```
Or
```suggestion
When this happens, the function pointers are cast to a generic function pointer type, and processed without knowledge of the type.
```
Either could work because they're both referring to the pointers accepted by the function, which are function pointers.